bvm/ballerina-core/src/main/java/org/ballerinalang/util/TableIterator.java

```diff
@@ -246,47 +253,136 @@ protected BNewArray getDataArray(Array array) throws SQLException {
+            }
+        } else if (firstNonNullElement instanceof Boolean) {
+            refValueArray = createEmptyRefValueArray(BTypes.typeBoolean, length);
+            for (int i = 0; i < length; i++) {
+                refValueArray.add(i, dataArray[i] != null ? new BBoolean((Boolean) dataArray[i]) : null);
+            }
+        } else if (firstNonNullElement instanceof Integer) {
+            refValueArray = createEmptyRefValueArray(BTypes.typeInt, length);
+            for (int i = 0; i < length; i++) {
+                refValueArray.add(i, dataArray[i] != null ? new BInteger((Integer) dataArray[i]) : null);
+            }
+        } else if (firstNonNullElement instanceof Long) {
+            refValueArray = createEmptyRefValueArray(BTypes.typeInt, length);
+            for (int i = 0; i < length; i++) {
+                refValueArray.add(i, dataArray[i] != null ? new BInteger((Long) dataArray[i]) : null);
+            }
+        } else if (firstNonNullElement instanceof Float) {
+            refValueArray = createEmptyRefValueArray(BTypes.typeFloat, length);
+            for (int i = 0; i < length; i++) {
+                refValueArray.add(i, dataArray[i] != null ? new BFloat((Float) dataArray[i]) : null);
+            }
+        } else if (firstNonNullElement instanceof Double) {
+            refValueArray = createEmptyRefValueArray(BTypes.typeFloat, length);
+            for (int i = 0; i < length; i++) {
+                refValueArray.add(i, dataArray[i] != null ? new BFloat((Double) dataArray[i]) : null);
+            }
         }
+        return refValueArray;
+    }
+
+    private BRefValueArray createEmptyRefValueArray(BType type, int length) {
+        List<BType> memberTypes = new ArrayList<>(2);
+        memberTypes.add(type);
+        memberTypes.add(BTypes.typeNull);
+        BUnionType unionType = new BUnionType(memberTypes);
+        return new BRefValueArray(new BRefType[length], unionType);
+    }
+
+    private ArrayElementNullabilityAttributes getArrayElementNullabilityInfo(Object[] objects) {
+        ArrayElementNullabilityAttributes arrayElementNullabilityAttributes = new ArrayElementNullabilityAttributes();
+        int i = 0;
+        while (i < objects.length) {
+            if (objects[i] != null) {
+                arrayElementNullabilityAttributes.setFirstNonNullElement(objects[i]);
+                if (i > 0) {
+                    arrayElementNullabilityAttributes.setContainsNull(true);
+                }
+                i++;
+                break;
+            }
+            i++;
+        }
+        if (!arrayElementNullabilityAttributes.containsNull()) {
+            while (i < objects.length) {
+                if (objects[i] == null) {
+                    arrayElementNullabilityAttributes.setContainsNull(true);
+                    break;
+                }
+                i++;
+            }
+        }
+        return arrayElementNullabilityAttributes;
```

Comment:
Shall we either add comments to explain the logic here or consider simplifying the logic?